File: exercises/day15-No-Exceptions-For-Failures/csharp/SantaChristmasList/Business.cs
namespace SantaChristmasList
{
    public class Business
    {
        private readonly IFactory _factory;
        private readonly IInventory _inventory;
        private readonly IWishList _wishList;

        public Business(IFactory factory, IInventory inventory, IWishList wishList)
        {
            _factory = factory;
            _inventory = inventory;
            _wishList = wishList;
        }

        public Sleigh LoadGiftsInSleigh(params Child[] children)
        {
            var sleigh = new Sleigh();

            foreach (var child in children)
            {
                try
                {
                    LoadGiftForChild(sleigh, child);
                }
                catch (Exception e) when (e is ChildWishNotFoundException
                                          or GiftNotManufacturedException
                                          or GiftOutOfStockException)
                {
                    throw new BusinessException("Unexpected error while loading sleigh", e);
                }
            }

            return sleigh;
        }

        private void LoadGiftForChild(Sleigh sleigh, Child child)
        {
            var gift = _wishList.IdentifyGift(child);
            if (gift is null)
                throw new ChildWishNotFoundException(child);

            var manufacturedGift = _factory.FindManufacturedGift(gift);
            if (manufacturedGift is null)
                throw new GiftNotManufacturedException(gift);

            var finalGift = _inventory.PickUpGift(manufacturedGift.BarCode);
            if (finalGift is null)
                throw new GiftOutOfStockException(manufacturedGift);

            sleigh.Put(child, $"Gift: {finalGift.Name} has been loaded!");
        }
    }
}

File: exercises/day15-No-Exceptions-For-Failures/scala/src/main/scala/santaChristmasList/operations/Business.scala
package santaChristmasList.operations

import santaChristmasList.operations.dependencies._
import santaChristmasList.operations.models._

class Business(factory: Factory, inventory: Inventory, wishList: WishList) {

  def loadGiftsInSleigh(children: Child*): Sleigh = {
    val sleigh = new Sleigh

    children.foreach { child =>
      try {
        loadGiftForChild(sleigh, child)
      } catch {
        case e @ (_: ChildWishNotFoundException
                | _: GiftNotManufacturedException
                | _: GiftOutOfStockException) =>
          throw new BusinessException("Unexpected error while loading sleigh", e)
      }
    }

    sleigh
  }

  private def loadGiftForChild(sleigh: Sleigh, child: Child): Unit = {
    val gift = Option(wishList.identifyGift(child))
      .getOrElse(throw new ChildWishNotFoundException(child))

    val manufacturedGift = Option(factory.findManufacturedGift(gift))
      .getOrElse(throw new GiftNotManufacturedException(gift))

    val finalGift = Option(inventory.pickUpGift(manufacturedGift.barCode))
      .getOrElse(throw new GiftOutOfStockException(manufacturedGift))

    sleigh.put(child, s"Gift: ${finalGift.name} has been loaded!")
  }
}

class ChildWishNotFoundException(child: Child)
  extends RuntimeException(s"No wish found for child: ${child.name}")

class GiftNotManufacturedException(gift: Gift)
  extends RuntimeException(s"Gift has not been manufactured: ${gift.name}")

class GiftOutOfStockException(gift: Gift)
  extends RuntimeException(s"Gift out of stock: ${gift.name}")

class BusinessException(message: String, cause: Throwable)
  extends RuntimeException(message, cause)
